### DISCUSSION TITLE
Web Progress Updates/Questions

### DISCUSSION THREAD

**Discussion by AKushWarrior** [isar/isar#23] [Ideas]
Figured I'd fork this off into a new discussion (since it's really not a subset of the Contribution one), so you can have a way to track what's currently accomplished.

## Current Goals in Isar-Web:
- Reimplementing IndexedDB (also referred to by me as IDB) APIs using a typed, promise-based, functional wrapper
    * Currently Done:
        - opening a database, opening a transaction, put, and get operations are supported & (semi-)tested.
    * TODO:
        - Work on finishing up porting CRUD operations
        - Implement getWhere and derivatives
        - Work on schema migrations (relatively difficult, not natively supported by IDB)
        - Figure out opening database semantics (kind of hacky rn, treating an Isar "database" as 1:1 with an IDB database)
- Creating a Dart interface to the aforementioned wrapper:
    * TODO:
        - Literally everything; can't start on this until the first thing is done.

## Questions:
- Is there a way you want me to handle the distinction between Object Stores and Databases?
    - What I mean by that: an "object store" is a schema-based store within a larger IDB database. An application can have infinite databases if it chooses; each database can have infinite object stores.
    - Right now, the wrapper is creating a new IDB database with one object store for each new Isar database. This is valid, and has pros and cons:
        - Opening a IDB database is expensive compared to opening an IDB object store, so loading time for an Isar instance will be relatively high. (**CON**)
        - Data independence is higher if we open a new database for each new Isar instance, and I believe that will help avoid sticky issues with schema migration. (**PRO**)
        - Relationships are largely impossible (or inefficient) if data is distributed between databases. We can likely solve this down the road (when relationships are supported) by putting related Isar instances as stores in the same database and separating them otherwise. (**CON**ish)
        - Data throughput is higher when running multiple transactions on different databases than multiple transactions on the same database. (**PRO**)
    - The other option is to only open one database, and put each Isar instance as a new object store in that database. The pros and cons for that are largely the inverse of the above. A notable trick we can pull off by putting multiple Isar instances as stores in the same database is creating transactions that span multiple stores, which makes cross-database techniques (relationships) way easier.

link to the repository for those interested: https://github.com/isar/isar-web

**Comment by simc** [comment]
Wow this sounds really good. I already thought a lot about web while implementing core.

### Databases vs Object stores

I would do it the following way:
- One database for each Isar instance, one object store for each collection
- Cross instance relationships don't really make sense because every instance has the same schema
- One "_isar" database that stores the schema. The reason is that you already need to have the schema when you open an instance database

### Schema migration
Schema migrations will be quite easy. We store JavaScript objects (e.g. Dart's `Map<String, dynamic>`) in the collections.

- When the user adds a field, we write it for future inserts, existing ones are `null`.
- When he deletes a field, it is just no longer written for future inserts and we ignore the field for already written objects.
- Changing the type of a field is not allowed and we will stop opening the instance
- Added and deleted collections (object stores) need to be handled
- Added and deleted (compound) indexes need to be handled

### Dart interface
The isar_generator generates a lot of stuff. In the future, it should generate:
- `isar.g.dart` file for platform independent code (schema, queries, static extensions etc.)
- `isar_web.g.dart` for an `isarOpen()` method and adapters to convert Dart objects to `Map<String, dynamic>`
- `isar_native.g.dart`for an `isarOpen()` method and adapters to convert Dart objects to the Isar binary representation

Here is an example schema that the generator creates:
```json
[
   {
      "name":"UserModel",
      "properties":[
         {
            "name":"name",
            "type":5,
            "isObjectId":true
         },
         {
            "name":"age",
            "type":3,
            "isObjectId":false
         },
         {
            "name":"friends",
            "type":11,
            "isObjectId":false
         }
      ],
      "indexes":[
         {
            "unique":false,
            "properties": [
               {
                  "name":"name",
                  "indexType":1,
                  "caseSensitive":true
               }
            ]
         }
      ]
   }
]
```

**Reply by AKushWarrior** [reply]
### Databases vs. Object Stores

I think I get what you're saying, but let me break it down conceptually to make sure:
- The user passes an optional String into the openIsar method; that's the database name. By default, it'll just be _isar or something.
- The schemas (collections) defined by the user are translated through some reflection/code generation process into object store definitions.
- The generator automagically (code diffs from previous generation?) detects migrations and handles versioning in IndexedDB.

### Schema Migration

This should be pretty simple, because IndexedDB has a built-in onUpgradeNeeded callback that will let us do migrations, where we can add/delete collections with no extra cost. 

Adding/deleting indexed fields is going to be more tricky (we'll need to manually migrate data, because IndexedDB will not automatically transfer data if the index schema of an object store is changed). Adding/deleting unindexed fields requires no extra work from us, beyond simply writing objects to the database.

### isar_web.g.dart

Sounds fine to me. This is actually probably the easiest part; with a well designed TypeScript API, this should just be a 1:1 mapping process. The most troubling part for me is figuring out when to trigger a migration.

**Reply by simc** [reply]
> The user passes an optional String into the openIsar method; that's the database name.

Exactly. I would use "isar" as default.

> The schemas (collections) defined by the user are translated through some reflection/code generation process into object store definitions.

Yes, the generator generates something like the schema json above.

> The generator automagically (code diffs from previous generation?) detects migrations and handles versioning in IndexedDB.

Not quite. I would store the schema in a "config" database (maybe call it "_isar") in indexeddb. You can then compare the current schema with the existing one to handle the necessary migrations.

> Adding/deleting indexed fields is going to be more tricky (we'll need to manually migrate data, because IndexedDB will not automatically transfer data if the index schema of an object store is changed). Adding/deleting unindexed fields requires no extra work from us, beyond simply writing objects to the database.

Not sure I really understand the problem. Why do we need to do manual migration? If we add a new index, won't indexeddb automatically populate it with the existing data?

**Comment by AKushWarrior** [comment]
@leisim I just found that Dart already has a wrapper for indexed_db as a core library maintained by the Dart team. https://api.dart.dev/stable/2.10.5/dart-indexed_db/dart-indexed_db-library.html

I'm essentially replicating this work in TypeScript right now anyways. Do you think it's feasible to just use the Dart library? Is there a significant performance benefit to doing it ourselves?

**Reply by simc** [reply]
I started with this library while developing Hive. There are two issues: 
1. It's not feature complete and very hard to extend or mix with native js/ts code
2. Dart - JS interfaces always come with some overhead and calls crossing between Dart and JS should be kept to a minimum

**Reply by tannermeade** [reply]
Thanks for the recent stable release to start the new year off -- so some good news with Dart / JS interop: in the recent Dart 2.15 update there were improvements to JS interop here (little to no overhead for some core usecases) https://github.com/dart-lang/sdk/blob/main/CHANGELOG.md#dartjs_util

This probably changes some of the considerations when it comes to building out the web support for isar.
